### PR TITLE
feat(mobile): Multiselect add to favorite from the timeline

### PR DESCRIPTION
**Pull request by martyfuhry**
Implements #1549. While multiselecting images, you can tap the "Favorite" star to add them to your favorites.

image

**Notes**
- Favorite is currently in between Share & Delete, but maybe it belongs first.
- Fixes and improves some of the logic and handling of adding images to favorites using the state providers
- Did I use the right star? I think it should be highlighted instead of the bordered star.

### CODE REVIEW

**Comment by EnochPrime** [comment]
Currently `StarOutline` is used in the web asset viewer when the assets is not a favorite and then `Star` when it is a favorite. The button toggles the state. 

From the photos page on web you can multiselect and select Add to Favorites, which has no icon. 

From the favorites page on web you can multiselect and remove from favorites which uses `StarMinusOutline`.

I would say use `StarPlus` to indicate the one way nature of adding. At least until someone works out what to do if you select some favorite and some not.

**Comment by martyfuhry** [comment]
> I would say use `StarPlus` to indicate the one way nature of adding. At least until someone works out what to do if you select some favorite and some not.

Hmm. Unfortunately, the default [icon set](https://fonts.google.com/icons?selected=Material+Icons&icon.set=Material+Icons) in Flutter doesn't seem to include anything that looks like a star with a plus sign on it. If you find something there you like, that's what we've been using elsewhere in the app. Anything else would have to be either designed or... I suppose I could draw something with a stack right in Flutter.

**Comment by martyfuhry** [comment]
Oh, and this PR doesn't really address what to do when some are favorited and some are not, nor does it address multi-selecting removing favorites from the favorites view. I can do that in this PR or a separate one.

**Comment by EnochPrime** [comment]
I checked the "reference app" and it apparently doesn't do multiselect favorites.

We could do, if selection is:
- Not favorited: display `StarOutline`, set all to favorite when tapped
- Favorited: display `Star`, set all to not favorite when tapped
- Mixed: display `StarHalf`, set not favorited to favorite when tapped

**Comment by matthinc** [comment]
Great Idea. Just one concern: You can multi-select local assets. These don't exist on the server so the api-call will result in a severe log warning. Also the favorite change would not persist across app restarts until we implement local favorites.

Idea: Disable the favorite button if one or more local assets are part of the selection.

File: mobile/lib/shared/providers/asset.provider.dart [hunk 1]
```diff
@@ -277,14 +277,12 @@ class AssetNotifier extends StateNotifier<AssetsState> {
...
-        }
-        return a;
-      }).toList(),
-    );
+    final index = state.allAssets.indexWhere((a) => asset.id == a.id);
+    if (index > 0) {
+      state.allAssets.removeAt(index);
+      state.allAssets.insert(index, Asset.remote(newAsset));
+      _updateAssetsState(state.allAssets);
+    }
```
Comment (matthinc) [inline mobile/lib/shared/providers/asset.provider.dart:285]: Great optimization!

Comment (martyfuhry) [reply mobile/lib/shared/providers/asset.provider.dart:285]: Thanks!

**Comment by martyfuhry** [comment]
> I checked the "reference app" and it apparently doesn't do multiselect favorites.
> 
> We could do, if selection is:
> 
>     * Not favorited: display `StarOutline`, set all to favorite when tapped
> 
>     * Favorited: display `Star`, set all to not favorite when tapped
> 
>     * Mixed: display `StarHalf`, set not favorited to favorite when tapped

That's super surprising. I thought for sure this would be a worthwhile feature. I suppose the "other app" folks didn't think so?

I like your ideas. For now, it just blanket applies a "favorite" to all the selected assets. There's no removal option (yet!). Do you want it in this PR or another one?

**Comment by martyfuhry** [comment]
> Great Idea. Just one concern: You can multi-select local assets. These don't exist on the server so the api-call will result in a severe log warning. Also the favorite change would not persist across app restarts until we implement local favorites.
> 
> Idea: Disable the favorite button if one or more local assets are part of the selection.

Thank you very much for the feedback. I'm still figuring out the mechanics of the local assets. For now, I applied the same filter as `onAddToAlbum` where we only select the remote assets.

**Review by alextran1502 — APPROVED** [review]
Look good! Feel free to merge when you are done

**Pull request merged**